File: src/main/java/org/veupathdb/service/eda/merge/core/request/MergedTabularRequestResources.java
package org.veupathdb.service.eda.ms.core.request;

import jakarta.ws.rs.BadRequestException;
import org.gusdb.fgputil.client.ResponseFuture;
import org.gusdb.fgputil.validation.ValidationException;
import org.veupathdb.service.eda.common.client.EdaComputeClient;
import org.veupathdb.service.eda.common.client.spec.StreamSpec;
import org.veupathdb.service.eda.generated.model.APIFilter;
import org.veupathdb.service.eda.generated.model.MergedEntityTabularPostRequest;
import org.veupathdb.service.eda.generated.model.VariableSpec;
import org.veupathdb.service.eda.ms.Resources;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map.Entry;
import java.util.Optional;

public class MergedTabularRequestResources extends RequestResources {

  private final EdaComputeClient _computeSvc;
  private final List<APIFilter> _subsetFilters;
  private final Optional<ComputeInfo> _computeInfo;
  private final String _targetEntityId;
  private final List<VariableSpec> _outputVarSpecs;

  public MergedTabularRequestResources(MergedEntityTabularPostRequest request, Entry<String, String> authHeader) throws ValidationException {
    super(request, authHeader);
    _targetEntityId = request.getEntityId();
    _outputVarSpecs = request.getOutputVariables();
    _computeSvc = new EdaComputeClient(Resources.COMPUTE_SERVICE_URL, authHeader);
    _subsetFilters = Optional.ofNullable(request.getFilters()).orElse(Collections.emptyList());
    _computeInfo = Optional.ofNullable(request.getComputeSpec())
        .map(spec -> new ComputeInfo(spec.getComputeName(),
            new EdaComputeClient.ComputeRequestBody(_metadata.getStudyId(), _subsetFilters, _derivedVariableSpecs, spec.getComputeConfig())));
    incorporateCompute();
  }

  private void incorporateCompute() {
    // if compute specified, check if compute results are available; throw if not, get computed metadata if so
    _computeInfo.ifPresent(info -> {
      if (!_computeSvc.isJobResultsAvailable(info.getComputeName(), info.getRequestBody()))
        throw new BadRequestException("Compute results are not available for the requested job.");
      else
        info.setMetadata(_computeSvc.getJobVariableMetadata(info.getComputeName(), info.getRequestBody()));
    });
    _computeInfo.ifPresent(computeInfo -> _metadata.incorporateComputedVariables(computeInfo.getVariables()));
  }

  public ResponseFuture getSubsettingTabularStream(StreamSpec spec) {

    // for derived var plugins, need to ensure filters overrides produce set of rows which are a subset of the rows
    //   produced by the "global" filters.  Easiest way to do that is to simply combine the filters, resulting in
    //   an intersection of the global subset and the overridden subset
    if (spec.getFiltersOverride().isPresent() && !_subsetFilters.isEmpty()) {
      StreamSpec modifiedSpec = new StreamSpec(spec.getStreamName(), spec.getEntityId());
      modifiedSpec.addAll(spec);
      List<APIFilter> combinedFilters = new ArrayList<>();
      combinedFilters.addAll(_subsetFilters);
      combinedFilters.addAll(spec.getFiltersOverride().get());
      modifiedSpec.setFiltersOverride(combinedFilters);
      spec = modifiedSpec;
    }

    return _subsetSvc.getTabularDataStream(_metadata, _subsetFilters, spec);
  }

  public ResponseFuture getComputeTabularStream() {
    return _computeSvc.getJobTabularOutput(_computeInfo.orElseThrow().getComputeName(), _computeInfo.get().getRequestBody());
  }

  public List<APIFilter> getSubsetFilters() { return _subsetFilters; }
  public Optional<ComputeInfo> getComputeInfo() { return _computeInfo; }
  public String getTargetEntityId() { return _targetEntityId; }
  public List<VariableSpec> getOutputVariableSpecs() { return _outputVarSpecs; }

}

File: src/main/kotlin/org/veupathdb/service/eda/merge/core/request/MergedTabularRequestResources.kt
package org.veupathdb.service.eda.merge.core.request

import jakarta.ws.rs.BadRequestException
import org.gusdb.fgputil.iterator.CloseableIterator
import org.gusdb.fgputil.validation.ValidationException
import org.veupathdb.service.eda.common.client.EdaComputeClient
import org.veupathdb.service.eda.common.client.EdaSubsettingClient
import org.veupathdb.service.eda.common.client.spec.EdaMergingSpecValidator
import org.veupathdb.service.eda.common.client.spec.StreamSpec
import org.veupathdb.service.eda.compute.EDACompute
import org.veupathdb.service.eda.compute.jobs.ReservedFiles
import org.veupathdb.service.eda.compute.plugins.PluginMeta
import org.veupathdb.service.eda.compute.plugins.PluginRegistry
import org.veupathdb.service.eda.generated.model.*

/**
 * Subclass of RequestResources which supplements the superclass' resources with
 * target entity, subset filters, and compute information needed for tabular
 * requests.  This includes fetching compute job metadata, computed variable
 * metadata (incorporated into the ReferenceMetadata instance), and a method
 * which provides the actual computed variable tabular stream, all by querying
 * the compute service.
 */
class MergedTabularRequestResources
@Throws(ValidationException::class)
constructor(
  studyId: String,
  variableSpecs: List<DerivedVariableSpec>,

  val subsetFilters: List<APIFilter>,
  val computeInfo: ComputeInfo?,
  val targetEntityId: String,
  val outputVariableSpecs: List<VariableSpec>,
) : RequestResources(studyId, variableSpecs) {
  @Deprecated("for removal")
  lateinit var computeSvc: EdaComputeClient

  constructor(request: MergedEntityTabularPostRequest) : this(
    request.studyId,
    request.derivedVariables ?: emptyList(),
    request.filters ?: emptyList(),
    request.computeSpec?.let { ComputeInfo(
      it.computeName,
      EdaComputeClient.ComputeRequestBody(
        request.studyId,
        request.filters ?: emptyList(),
        request.derivedVariables ?: emptyList(),
        it.computeConfig
      )
    ) },
    request.entityId,
    request.outputVariables ?: emptyList(),
  )

  init {
    // incorporate computed metadata (if compute info present)
    incorporateCompute()

    // validation of incoming request, based specifically on the requested
    // entity done during spec creation
    validateIncomingRequest()
  }

  fun getSubsettingTabularStream(spec: StreamSpec): CloseableIterator<Map<String, String>> {
    // for derived var plugins, need to ensure filters overrides produce set of
    // rows which are a subset of the rows produced by the "global" filters.
    // Easiest way to do that is to simply combine the filters, resulting in an
    // intersection of the global subset and the overridden subset.
    return if (spec.filtersOverride.isPresent && subsetFilters.isNotEmpty()) {
      StreamSpec(spec.streamName, spec.entityId).apply {
        addAll(spec)
        setFiltersOverride(subsetFilters + spec.filtersOverride.get())
      }.let { EdaSubsettingClient.getTabularDataIterator(metadata.studyId, subsetFilters, it) }
    } else {
      EdaSubsettingClient.getTabularDataIterator(metadata.studyId, subsetFilters, spec)
    }
  }

  fun getComputeStreamIterator(): CloseableIterator<Map<String, String>> {
    if (computeInfo == null)
      throw IllegalStateException("Cannot get compute stream iterator if no compute is specified in request.")

    val file = EDACompute.getComputeJobFiles(
      PluginRegistry.get(computeInfo.computeName) as PluginMeta<out ComputeRequestBase>,
      computeInfo.requestBody,
    ).first { it.name == ReservedFiles.OutputTabular }

    return EdaComputeClient.getJobTabularIteratorOutput(
      metadata.getEntity(computeInfo.computeEntity).orElseThrow(),
      computeInfo.variables,
      metadata,
      file::open
    )
  }

  private fun incorporateCompute() {
    // if compute specified, check if compute results are available; throw if
    // not, get computed metadata if so
    computeInfo?.also {
      if (!computeSvc.isJobResultsAvailable(it.computeName, it.requestBody))
        throw BadRequestException("Compute results are not available for the requested job.")
      else
        it.setMetadata(computeSvc.getJobVariableMetadata(it.computeName, it.requestBody))

      metadata.incorporateComputedVariables(it.variables)
    }
  }

  @Throws(ValidationException::class)
  private fun validateIncomingRequest() {
    // create a stream spec from the request input and validate using merge svc
    // spec validator
    val requestSpec = StreamSpec("incoming", targetEntityId)
    requestSpec.addAll(outputVariableSpecs)

    EdaMergingSpecValidator()
      .validateStreamSpecs(listOf(requestSpec), metadata)
      .throwIfInvalid()

    // if compute was requested, make sure the computed entity is the same as,
    // or an ancestor of, the target entity of this request
    if (computeInfo != null) {
      if (
        targetEntityId != computeInfo.computeEntity
        && metadata.getAncestors(metadata.getEntity(targetEntityId).orElseThrow())
          .none { it.id == computeInfo.computeEntity }
      ) {
        // we don't perform reductions on computed vars so they must be on the target entity or an ancestor
        throw ValidationException("Entity of computed variable must be the same as, or ancestor of, the target entity")
      }
    }
  }
}
